### ISSUE TITLE
Show NON_FIELD_ERRORS when page does not pass validation

### ISSUE THREAD

**Issue by thenewguy** [wagtail/wagtail#536]
I am not using the latest version yet, so this may already be fixed, but...

I am trying to use the Model.clean() method on a page type to validate a few conditional fields that depend on another model field.

Along the lines of if page "mode" is 1, fields "foo" and "bar" should not be 3 or 5.

When I raise a validation error the dashboard shows "The page could not be saved due to validation errors" but it doesn't output the error messages.

How are other people doing this type of validation now?

**Comment by thenewguy** [comment]
This is still an issue on Wagtail 1.1.  If you have a snippet with a unique together constraint like 
```
    class Meta:
        unique_together = ("position", "primary")
```
you just get an red error banner at the top that says "The snippet could not be created due to errors." without mentioning the underlying problem so an ordinary user has no way to correct the mistakes.

Django handles these errors natively and exposes a simple interface to output them in a template.  It is hacky to override model validation to assign the errors to specific fields, and since they pertain to multiple fields the same error message needs to be displayed under each field which isn't particularly clean.

**Comment by thenewguy** [comment]
Plus you basically have to redo the validation to determine which fields are invalid.  In the case of unique fields you also have to redo all the database queries to determine which unique constraints are being violated.

**Comment by thenewguy** [comment]
The following hack is the best way I've come up with to deal with unique_together specifically:
```
    def _perform_unique_checks(self, unique_checks):
        """
            HACK TO DEAL WITH WAGTAIL IGNORING NON_FIELD_ERRORS
            
            https://github.com/torchbox/wagtail/issues/536
        """
        errors = {}
        for unique_check in unique_checks:
            local_errors = super(ExternalPanelSnippet, self)._perform_unique_checks([unique_check])
            
            for key, errs in local_errors.items():
                if key == NON_FIELD_ERRORS:
                    for field_name in unique_check[1]:
                        errors.setdefault(field_name, []).extend(errs)
                else:
                    errors.setdefault(key, []).extend(errs)
        
        return errors
```

But it requires overriding private methods in Django model validation.

Is supporting NON_FIELD_ERRORS still rejected?

Can you explain what you mean by this following statement?

>At present if the user makes 5 content entry mistakes, 4 of which are on specific fields and 1 of 
>which is the example you've provided, then tries to save, the error at the top of the screen (the 
>banner that animates down from the top) will only refer to that non_field_error. 
>
>This is a UX issue - despite making 5 mistakes, only 1 is apparently mentioned. Indeed if the user 
>fixes the 1 error and saves again, they're then told about the other 4, which is terribly frustrating and 
>not something we want to occur in Wagtail.

Why does displaying NON_FIELD_ERRORS in addition to field specific errors result in the case you've described?

**Comment by thenewguy** [comment]
I have encountered another issue where NON_FIELD_ERRORS is needed.  StreamField does not seem to output errors assigned to it.

This is important to allow an editor to use free form or rich text content on demand as needed.  So in the clean method you would check to verify one of the content fields has a value but not both.

Say you have the following snippet:
```
@register_snippet
@python_2_unicode_compatible
class SidebarSnippet(models.Model):
    
	...
	
	free_form_content = StreamField(FreeFormBlock, blank=True)
    rich_text_content = RichTextField(blank=True)
    
	...
	
    panels = [
        ...
        
		StreamFieldPanel("free_form_content"),
        RichTextFieldPanel("rich_text_content"),
		
		...
    ]
	
	def clean(self):
        errors = {}
        
        content_fields = ("free_form_content", "rich_text_content")
        content_values = [getattr(self, attr) for attr in content_fields]
        
        if all(content_values):
            msg = "You cannot specify %s together." % " and ".join(content_fields)
            for field_name in content_fields:
                errors.setdefault(field_name, []).append(msg)
        
        elif not any(content_values):
            msg = "You must specify one of %s" % " or ".join(content_fields)
            for field_name in content_fields:
                errors.setdefault(field_name, []).append(msg)
        
        if errors:
            raise ValidationError(errors)
```

The error is only presented for `rich_text_content`.

This shows the error cannot be shown for `free_form_content` at all (according to the documentation I think this is expected?):
```
    def clean(self):
        raise ValidationError({"free_form_content": "error here"})
```

**Comment by Croden** [comment]
To output non_field_errors from forms, I overwritten message part in wagtailadmin/base.html. Working properly with Snippets and Page models:

```
...
{% if messages or edit_handler and edit_handler.form and edit_handler.form.non_field_errors %}
    <ul>
        {% for message in messages %}
            <li{% if message.tags %} class="{{ message.tags }}"{% endif %}>{{ message|safe }}</li>
        {% endfor %}
        {% if edit_handler and edit_handler.form and edit_handler.form.non_field_errors %}
            <li class="error">{{ edit_handler.form.non_field_errors.as_text }}</li>
        {% endif %}
    </ul>
{% endif %}
...
```

**Comment by rezhajulio** [comment]
@Croden Thanks!

**Comment by shortly-portly** [comment]
Hi.

Is there any reason why @Croden solution couldn't be incorporated into Wagtail itself?

**Comment by gasman** [comment]
Being addressed by #2713 and #2783

**Comment by gasman** [comment]
Have (finally!) reviewed #2713 and #2783, which unfortunately have various structural and UX issues between them (they involve wagtailadmin/base.html digging into template context to look for form errors, rather than passing that information via the messages framework; as @davecranwell points out above, highlighting _only_ non-field errors in the message means that any field errors lose prominence; and they're missing tests).

I think we can handle this neatly by introducing a new `validation_error` helper to wagtailadmin.messages - this would use the same styling as the 'error' helper, but would receive the failing form as a parameter and deal with formatting the form errors in the most UX-friendly way (e.g. showing the full set of field and non-field errors whenever a non-field error exists; but if only field errors are present, just show the generic "there were validation errors" message and leave the per-field highlighting to do the rest). This also makes it easy to reuse the logic on other admin forms (such as snippets / modeladmin). I'll put together a PR for this now.

**Comment by bufke** [comment]
Should this be fixed in 1.10 for snippets? I'm still seeing the problem. ValidationError in a model's clean function just gives me "The snippet could not be saved due to errors." in the edit snippet interface.

**Comment by ababic** [comment]
Hi @bufke. Not yet. #3441 exists to cover that, but it's yet to be implemented.
